### ISSUE TITLE
Turn msg.value in non-payable function or modifier into error

### ISSUE THREAD

**Issue by erak** [ethereum/solidity#4601]
After fixing https://github.com/ethereum/solidity/issues/3770, we should discuss if the usage of ``msg.value`` in non-payble functions and modifiers can be turned into an error:
```
contract C {
    modifier costs(uint _amount) { require(msg.value >= _amount); _; }
    function f(uint _amount) public { require(msg.value >= _amount); }
}
```
Output (after PR https://github.com/ethereum/solidity/pull/4590 got merged):
```
/tmp/x.sol:3:62: Warning: "msg.value" used in non-payable function. Do you want to add the "payable" modifier to this function?
    function f(uint _amount) public costs(1 ether) { require(msg.value >= _amount); }
                                                             ^-------^
/tmp/x.sol:2:44: Warning: Modifier used in non-payable function, but this expression requires the function to be payable.
    modifier costs(uint _amount) { require(msg.value >= _amount); _; }
                                           ^-------^
/tmp/x.sol:3:37: The modifier is used here:
    function f(uint _amount) public costs(1 ether) { require(msg.value >= _amount); }
                                    ^------------^

```

**Comment by axic** [comment]
I think the reason behind this might have been that the same modifier can be used by payable and non-payable functions.

Is there any reason to keep this? Can we turn the warning into an error?

**Comment by axic** [comment]
@chriseth any opinion on this?

**Comment by chriseth** [comment]
We seem to have decided that during the meeting. So error it is?

**Comment by chriseth** [comment]
FIxed in #4590
